src/circuit/circuit_frame.rs

```diff
@@ -4644,15 +4643,117 @@ pub fn comparison_helper<F: LurkField, CS: ConstraintSystem<F>>(
     Ok(())
 }
 
+#[derive(Copy, Clone)]
+pub enum UnsignedType {
+    U32,
+    U64,
+}
+
+// Convert from num to u32.
+pub fn to_unsigned_int<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    maybe_unsigned: &AllocatedNum<F>,
+    size: UnsignedType,
+) -> Result<AllocatedNum<F>, SynthesisError> {
+
+    let p_bn =  match size {
+        UnsignedType::U32 => BigUint::pow(&BigUint::from_u32(2).unwrap(), 32),
+        UnsignedType::U64 => BigUint::pow(&BigUint::from_u32(2).unwrap(), 64),
+    };
+
+    let v = match maybe_unsigned.get_value() {
+        Some(v) => v,
+        None => F::zero(), //dummy
+    };
+    let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
+
+    let (q_bn, r_bn) = field_bn.div_rem(&p_bn);
+
+    let q_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "q"), q_bn)?;
+    let r_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "r"), r_bn)?;
+
+    // a = b.q + r
+    let product = match size {
+        UnsignedType::U32 =>  mul(
+            &mut cs.namespace(|| "product(q,pow(2,32))"),
+            &q_num,
+            &g.power2_32_num,
+        )?,
+        UnsignedType::U64 =>  mul(
+            &mut cs.namespace(|| "product(q,pow(2,64))"),
+            &q_num,
+            &g.power2_64_num,
+        )?,
+    };
+    let sum = add(&mut cs.namespace(|| "sum remainder"), &product, &r_num)?;
+    let u_decomp = alloc_equal(
+        &mut cs.namespace(|| "check unsigned decomposition"),
+        &sum,
+        maybe_unsigned,
+    )?;
+    enforce_true(
+        &mut cs.namespace(|| "enforce decomposition"),
+        &u_decomp,
+    )?;
+
+    // enforce remainder range
+    let r_bits = r_num.to_bits_le(&mut cs.namespace(|| "u64 remainder bit decomp"))?;
+    let output = match size {
+        UnsignedType::U32 => {
+            enforce_at_most_n_bits(
+                &mut cs.namespace(|| "remainder u32 range"),
+                g,
+                r_bits,
+                32,
+            )?;
+            let output_alloc = AllocatedNum::alloc(&mut cs.namespace(|| "to_32"), || {
+                Ok(F::from_u32(v.to_u32_unchecked()).unwrap())
+            })?;
+            let u32_is_remainder = alloc_equal(
+                &mut cs.namespace(|| "check u32 value is the remainder"),
+                &r_num,
+                &output_alloc,
+            )?;
+            enforce_true(
+                &mut cs.namespace(|| "enforce u32 decomposition output"),
+                &u32_is_remainder,
+            )?;
+            output_alloc
+        },
+        UnsignedType::U64 => {
+            enforce_at_most_n_bits(
+                &mut cs.namespace(|| "remainder u64 range"),
+                g,
+                r_bits,
+                64,
+            )?;
+            let output_alloc = AllocatedNum::alloc(&mut cs.namespace(|| "to_64"), || {
+                Ok(F::from_u64(v.to_u64_unchecked()).unwrap())
```

Comment:
Same question as `Ok(F::from_u32(v.to_u32_unchecked()).unwrap())` above on the `unwrap()`.

Reply:
`to_u32_unchecked()` and `to_u64_unchecked` return the last 32 or 64 bits of the field element. If the field element is larger than MAX_U32, then we just through away the most significant bits. In the fields used by Lurk, the modulus are never smaller than MAX_U32 or MAX_U64.

Reply:
> to_u32_unchecked() and to_u64_unchecked return the last 32 or 64 bits of the field element. If the field element is larger than MAX_U32, then we just through away the most significant bits. 

Right. No issue there.

> In the fields used by Lurk, the modulus are never smaller than MAX_U32 or MAX_U64.

Should we then at least document that `from_u{32,64}` are expected to panic when the modulus is smaller than MAX_U{32,64} in the `LurkField` trait then, for future implementations?

The Goldilocks field has modulus 2^64 - 2^32 + 1 < MAX_U64, and has some usage in our ecosystem ([some refs here](https://github.com/ingonyama-zk/papers/blob/main/goldilocks_ntt_trick.pdf) ).

Reply:
A couple thoughts. Documentation can't hurt, but it's worth noting that assumptions about fields size are implicitly documented since the representations are assumed to be 32 bytes. We might add support for other fields, and if larger this would be a relatively small change. It would require significantly more thought to consider the implications and needed to design changes to use very small fields — since we depend on field size for hash security (critical for Lurk soundness).

Since `LurkField::from_bytes()` will already panic (in `copy_from_slice`) if the wrong number of bytes are supplied, I think that serves as an effective guard against someone accidentally rushing forward unaware of the issues. We could put an explicit assertion on the length of the bytes supplied there as well, to more explicitly document that this is an intended failure. If/when we explicitly support other fields (like BLS12-377), we can stop to think through how to document and enforce an appropriate set of invariants. For now, Lurk assumes ~256-bit fields.

Reply:
> it's worth noting that assumptions about fields size are implicitly documented since the representations are assumed to be 32 bytes [...] Since LurkField::from_bytes() will already panic (in copy_from_slice) if the wrong number of bytes are supplied, I think that serves as an effective guard against someone accidentally rushing forward unaware of the issues.

That's definitely correct for the current instances of `LurkField`, but is it really impossible to write an implementation of `LurkField` with a `<Self as PrimeField>::Repr` that uses less than 32 bytes?

However, panicking with when receiving a different byte length than the representation demands isn't what I was thinking of. Rather, it's that (as you know) not all correctly-sized byte representations encode canonical (reduced) field elements. 

But I think I just missed that `AllocatedNum::to_repr()` can only produce canonical representations, i.e. the word `standard` is load-bearing here:
```
    /// Converts an element of the prime field into the standard byte representation for
    /// this field.
```
https://github.com/zkcrypto/ff/blob/e853770f4843e47cd5e07330140b5bd1cdfc9f11/src/lib.rs#L287-L288

My bad, nothing to see here. Plus with the most recent refactor, byte-to-field conversions are avoided entirely.

>  It would require significantly more thought to consider the implications and needed to design changes to use very 
small fields — since we depend on field size for hash security (critical for Lurk soundness).

Fair enough - indeed the Goldilocks field is used as a building block of field extensions in that context.

Reply:
> That's definitely correct for the current instances of LurkField, but is it really impossible to write an implementation of LurkField with a <Self as PrimeField>::Repr that uses less than 32 bytes?

No, it's not.

> But I think I just missed that PrimeField::to_repr() can only produce canonical representations, i.e. the word standard is load-bearing here:
>
>    /// Converts an element of the prime field into the standard byte representation for
>   /// this field.
> https://github.com/zkcrypto/ff/blob/e853770f4843e47cd5e07330140b5bd1cdfc9f11/src/lib.rs#L287-L288


I think 'standard' here is referring to the endianness. Also, I think you quoted the description of `to_repr`. `from_repr` does explain that only a canonical representation will successfully yield a field element. (I think you figured this part out. I'm just adding the reference here for discussion context.)

    /// Attempts to convert a byte representation of a field element into an element of
    /// this prime field, failing if the input is not canonical (is not smaller than the
    /// field's modulus).

https://github.com/zkcrypto/ff/blob/e853770f4843e47cd5e07330140b5bd1cdfc9f11/src/lib.rs#L264-L266